src/pulsar_producer.erl

```diff
@@ -279,3 +431,185 @@ next_sequence_id(State = #state{sequence_id = SequenceId}) ->
     State#state{sequence_id = SequenceId+1}.
 
 log_error(Fmt, Args) -> logger:error("[pulsar-producer] " ++ Fmt, Args).
+
+invoke_callback(Callback, Resp) ->
+    invoke_callback(Callback, Resp, _BatchLen = 1).
+
+invoke_callback(_Callback = undefined, _Resp, _BatchLen) ->
+    ok;
+invoke_callback({M, F, A}, Resp, BatchLen) ->
+    lists:foreach(
+      fun(_) ->
+        erlang:apply(M, F, [Resp] ++ A)
+      end,  lists:seq(1, BatchLen));
+invoke_callback(Callback, Resp, BatchLen) when is_function(Callback, 1) ->
+    lists:foreach(
+      fun(_) ->
+        Callback(Resp)
+      end,  lists:seq(1, BatchLen)).
+
+queue_item_sizer(?Q_ITEM(_CallId, _Ts, _Batch) = Item) ->
+    size(queue_item_marshaller(Item)).
+
+queue_item_marshaller(?Q_ITEM(_, _, _) = I) ->
+  term_to_binary(I);
+queue_item_marshaller(Bin) when is_binary(Bin) ->
+  binary_to_term(Bin).
+
+now_ts() ->
+    erlang:system_time(millisecond).
+
+make_queue_item(From, Messages) ->
+    ?Q_ITEM(From, now_ts(), Messages).
+
+enqueue_send_requests(Requests, State = #state{opts = #{replayq := Q} = Opts0}) ->
+    QItems = lists:map(
+               fun(?SEND_REQ(From, Messages)) ->
+                 make_queue_item(From, Messages)
+               end,
+               Requests),
+    NewQ = replayq:append(Q, QItems),
+    State#state{opts = Opts0#{replayq := NewQ}}.
```

Comment:
feels wrong having replayq deep in the opts.
should we consider changing `#state{}` to a map ?

Reply:
I put it there exactly to avoid messing with the state record during relup  :joy_cat: 

I'll migrate it to a map then.